### ISSUE TITLE
Follow up on #2123 (SYNC/ASYNC/UNKNOWN stuff)

### ISSUE THREAD

**Issue by bsideup** [reactor/reactor-core#2242]
Quoting @simonbasle, the current approach is focused around a `DIRECTLY_INTODUCES_A_THREAD_BOUNDARY` attribute with the following options:
- `NO WAY TO KNOW FOR SURE`
- `IT COULD`
- `NO`

I suggest we step back a little and think about our primary use case: "being able to know when a certain operator is synchronous, so that there is no need to instrument it (e.g. for ThreadLocals propagation)".

For that, we could use a marker (`SYNC`), not an attribute, which will hint that given `Scannable` is fully synchronous.

The absence of it would mean either "synchronous" or "unknown". Now the question is, do we need to distinguish these two?
We obviously don't need it for our primary use care, but I tried to think of other use cases (visualization, debugging, etc etc), yet I haven't found a use case where binary (sync/not sync) marker won't be sufficient.

The good thing is that if we make a decision today to remove `ASYNC` & `UNKNOWN`, we will be able to add them later.

But I think it'd be better to simplify it for now to avoid a confusion that we have right now. Some examples:
- `FluxConcatMap`/`FluxFlatMap` are `SYNC`, although these are the most used operators for asynchronous operations. The assumptions is that the `Publisher` returned from `mapper` would determine the async nature, but it is unclear.
- At the same time, `FluxDelaySequence` is `ASYNC`, although `FluxDelaySequence` can be seen as an optimized version of `concatMap(it -> Mono.delay(d).thenReturn(it))`, which will be marked as `SYNC` when you inspect the result of `concatMap`.
- `FluxWindowBoundary` is `SYNC` although it emits on timer's completion and, in fact, changes the threads (could also be a bug)

**Comment by simonbasle** [comment]
Again, this whole debate makes me think that `SYNC` and `ASYNC` are not the correct terms (which are maybe too overloaded). Maybe having the `Publisher` themselves answer to the scan was a bad idea in the end? Sleuth manipulates `Subscriber` anyway, IIRC, so it could still filter if the attribute was only scannable at subscriber level.

**Comment by simonbasle** [comment]
@bsideup @aneveu since this hasn't been released in anything else than snapshots yet, what do you think about the following renaming:

 - [ ] `Attr.RUN_STYLE` -> `Attr.THREAD_MODEL`
 - [ ] `RunStyle` -> `ThreadModel`
 - [ ] `RunStyle.SYNC` -> `ThreadModel.CONTINUE_ON_SAME`
 - [ ] `RunStyle.ASYNC` -> `ThreadModel.SCHEDULE_TASKS`
 - [ ] `RunStyle.UNKNOWN` -> `ThreadModel.UNKNOWN`

Also, the attribute should be documented to reflect the slight difference of meaning when queried on a `Publisher` vs a `Subscriber`.

I went back to the original issue and having the attribute on `Publisher` level opens more possibilities of better filtering, since otherwise the subscribers wrapping could be avoided but we'd still wrap publishers, for nothing.

For Publisher level operators like `repeatWhen` or `flatMap` that don't directly use a `Scheduler`, but could end up producing signals on a different thread due to the inner companion flux/monos, we could add another intermediate value `ThreadModel.DELEGATE_TO_COMPANION` ?
I don't think that value would make sense on the `Subscriber`s though.

**Comment by bsideup** [comment]
@simonbasle I still don't see a value in such precision and insist on a _marker_ approach instead, but I will let the end user / requester to decide.

**Comment by simonbasle** [comment]
pinging @robotmrv

**Comment by robotmrv** [comment]
@simonbasle  @bsideup  
sorry for being quiet for while (do not have much time at the moment)  

Naming for me is not very important, so I will leave this decision to you.    

Regarding granularity.  
Initial request (https://github.com/reactor/reactor-core/issues/2058) was about just a flag (true/false).  
So for me it would be fine to have ability just to know whether it could change `Thread` or not.  

> Attr.RUN_STYLE -> Attr.THREAD_MODEL
 RunStyle -> ThreadModel
 RunStyle.SYNC -> ThreadModel.CONTINUE_ON_SAME
 RunStyle.ASYNC -> ThreadModel.SCHEDULE_TASKS
 RunStyle.UNKNOWN -> ThreadModel.UNKNOWN

I am not sure that clients want to play with all this values.     
But I think only about `ThreadLocal`s case.  

Regarding 
>For Publisher level operators like repeatWhen or flatMap that don't directly use a Scheduler, but could end up producing signals on a different thread due to the inner companion flux/monos, we could add another intermediate value ThreadModel.DELEGATE_TO_COMPANION

Actually I thought (for simplicity, because of non `reactor-core` publishers or `Processor`s which may not apply onEachOpearator hook) that all operators which could produce (or consume) `Publisher` would be marked as "could potentially change Thread", but you decided to make more precise marking which allows even more efficient decorating.

**Comment by robotmrv** [comment]
Hi @simonbasle @bsideup   
I've found a time to experiment with solution from #2123 on Sleuth code base  
Implementation could be found [here](https://github.com/robotmrv/spring-cloud-sleuth/commit/7a2e9b08f386b8a9d4847beb59e2b221e21b155e#diff-db43984e17bb56450de433339b89f96fR188)  

There are some positive results:  
for `onEach` instrumentation allocation rate reduced from +50% to +27.5 (almost the same as `onLast` instrumentation +26.5%) comparing to base without instrumentation  
data is collected based on [Sleuth benchmark](https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/benchmarks/src/test/java/org/springframework/cloud/sleuth/benchmarks/jmh/webflux/MicroBenchmarkHttpTests.java)  

But there are several problems:  
Current approach requires traverse upstream and check whether instrumentation was applied, but 
1. It can't be done using `Operators.liftPublisher()` as there is no any marker that can be used as identifier on `Publisher` level 
even if use Lift Publisher's name as identifier (looks weird and wrong) there are could be more than one `onEach` hooks enabled so need to have access to the original lifter function (or other way to get related hook name on assembly time) to determine that it is your `Lifter`  
2. it can be work arounded by creating own `Lift Publisher` but it is mostly copy paste from `reactor-core` that looks wrong as core should allow add such `lifters` without such hacks.  
Current POC uses this approach, but it is not possible to cover `ParralleFlux` case as I faced the same issue as in #2328  

As the solution to this problems it could be used proposal from original issue #2058 (from second part):  
> add new Scannable attribute like
> 
> `Attr<Scannable> LIFTER = new Attr<>(null, Scannable::from)`  
> to all Lifter Publishers
> Returned Scannable should have such attributes
> `reactor.core.Scannable.Attr#NAME - by default toString()`  
> and
> `reactor.core.Scannable.Attr#ACTUAL - actual lifter function. probably could be helpful`

It could different names etc but main point is - access to the original lifter
WDYT?
